docs/proposals/20200720-single-controller-multitenancy.md

```diff
@@ -0,0 +1,548 @@
+    namespace: capz-system
+  allowedNamespaces:
+  - "test"
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: "test-account-creds"
+  namespace: capz-system
+stringData:
+ tenantID: 6bec3eaa-cac6-11ea-87d0-0242ac130003
+```
+
+Example: AzureUserAssignedIdentity
+
+```yaml
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: AzureCluster
+metadata:
+  name: "test"
+  namespace: "test"
+spec:
+  region: "westus2"
+  principalRef:
+    kind: AzureUserAssignedIdentity
+    name: test-account
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: AzureUserAssignedIdentity
+metadata:
+  name: "test-account"
+spec:
+  secretRef:
+    name: test-account-creds
+    namespace: capz-system
+  allowedNamespaces:
+  - "test"
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: "test-account-creds"
+  namespace: capz-system
+stringData:
+ tenantID: 6bec3eaa-cac6-11ea-87d0-0242ac130003
+ clientID: 74e870e6-cac6-11ea-87d0-0242ac130003
+```
+
+### Controller Changes
+
+* If principalRef is specified, the CRD is fetched and unmarshalled into a Azure `Authorizer` for 
+  the principal type.
+* The controller will compare the hash of the credential provider against the same secret’s provider
+  in a cache ([NFR2](#NFR2)).
+* The controller will take the newer of the two and instantiate Azure clients with the selected 
+  `Authorizer`.
+* The controller will set the Identity resource as one of the OwnerReferences of the AzureCluster.
+
+### Clusterctl changes
+
+Today, `clusterctl move` operates by tracking objectreferences within the same namespace, since we 
+are now proposing to use cluster-scoped resources, we will need to add requisite support to 
+clusterctl's object graph to track ownerReferences pointing at cluster-scoped resources, and ensure 
+they are moved. We will naively not delete cluster-scoped resources during a move, as they maybe 
+referenced across namespaces.
+
+#### Validating webhook changes
+
+None.
+
+A validating webhook could potentially handle some cross-resource validation necessary for the [security
+model](#security-model) and provide more immediate feedback to end users. However, it would be 
+imperfect. For example, a change to an Azure Identity resource could affect the validity of 
+corresponding AzureCluster.
+
+#### Principal Type Credential Provider Behaviour
+
+Implementations for all principal types will implement the AutoREST `Authorizer` interface as well 
+as an additional function signature to support caching.
+
+```go
+// Authorizer is the interface that provides a PrepareDecorator used to supply request
+// authorization. Most often, the Authorizer decorator runs last so it has access to the full
+// state of the formed HTTP request.
+type Authorizer interface {
+    WithAuthorization() PrepareDecorator
+}
+
+type AzureIdentityProvider interface {
+    Authorizer
+
+    // Hash returns a unique hash of the data forming the credentials
+    // for this principal
+    Hash() (string, error)
```

Comment:
watch for changes on observedGeneration to the secret for same effect?

Reply:
nvm, it's an interface so needs to be portable. You could treat it like a monotonic counter (because it is) but I can see why a hash is more elegant here

Reply:
👍